TestOperatorStats.testAddMixedStartingWithMergeable is flaky

Example failure: https://github.com/prestodb/presto/actions/runs/29297261905/job/86973406288?pr=28083

Error stack:

```
Error:  Tests run: 9745, Failures: 6, Errors: 0, Skipped: 1, Time elapsed: 775.016 s <<< FAILURE! - in TestSuite
Error:  com.facebook.presto.operator.TestOperatorStats.testAddMixedStartingWithNonMergeable  Time elapsed: 0.001 s  <<< FAILURE!
java.lang.AssertionError: Sets differ: expected [1, 2, 3] but got [1, 2, 3]
	at org.testng.Assert.fail(Assert.java:110)
	at org.testng.Assert.assertEquals(Assert.java:1825)
	at org.testng.Assert.assertEquals(Assert.java:1752)
	at com.facebook.presto.operator.TestOperatorStats.testAddMixedStartingWithNonMergeable(TestOperatorStats.java:388)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)

Error:  com.facebook.presto.operator.TestOperatorStats.testAddMixedStartingWithMergeable  Time elapsed: 0.001 s  <<< FAILURE!
java.lang.AssertionError: Sets differ: expected [1, 2, 3] but got [1, 2, 3]
	at org.testng.Assert.fail(Assert.java:110)
	at org.testng.Assert.assertEquals(Assert.java:1825)
	at org.testng.Assert.assertEquals(Assert.java:1752)
	at com.facebook.presto.operator.TestOperatorStats.testAddMixedStartingWithMergeable(TestOperatorStats.java:282)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)

Error:  com.facebook.presto.operator.TestOperatorStats.testAddNonMergeable  Time elapsed: 0 s  <<< FAILURE!
java.lang.AssertionError: Sets differ: expected [1, 2, 3] but got [1, 2, 3]
	at org.testng.Assert.fail(Assert.java:110)
	at org.testng.Assert.assertEquals(Assert.java:1825)
	at org.testng.Assert.assertEquals(Assert.java:1752)
	at com.facebook.presto.operator.TestOperatorStats.testAddNonMergeable(TestOperatorStats.java:438)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)

Error:  com.facebook.presto.operator.TestOperatorStats.testJson  Time elapsed: 0 s  <<< FAILURE!
java.lang.AssertionError: Sets differ: expected [1, 2, 3] but got [1, 2, 3]
	at org.testng.Assert.fail(Assert.java:110)
	at org.testng.Assert.assertEquals(Assert.java:1825)
	at org.testng.Assert.assertEquals(Assert.java:1752)
	at com.facebook.presto.operator.TestOperatorStats.assertExpectedOperatorStats(TestOperatorStats.java:228)
	at com.facebook.presto.operator.TestOperatorStats.testJson(TestOperatorStats.java:173)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)

Error:  com.facebook.presto.operator.TestTaskStats.testJson  Time elapsed: 0.002 s  <<< FAILURE!
java.lang.AssertionError: Sets differ: expected [1, 2, 3] but got [1, 2, 3]
	at org.testng.Assert.fail(Assert.java:110)
	at org.testng.Assert.assertEquals(Assert.java:1825)
	at org.testng.Assert.assertEquals(Assert.java:1752)
	at com.facebook.presto.operator.TestOperatorStats.assertExpectedOperatorStats(TestOperatorStats.java:228)
	at com.facebook.presto.operator.TestPipelineStats.assertExpectedPipelineStats(TestPipelineStats.java:134)
	at com.facebook.presto.operator.TestTaskStats.assertExpectedTaskStats(TestTaskStats.java:144)
	at com.facebook.presto.operator.TestTaskStats.testJson(TestTaskStats.java:97)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)

Error:  com.facebook.presto.operator.TestDriverStats.testJson  Time elapsed: 0.001 s  <<< FAILURE!
java.lang.AssertionError: Sets differ: expected [1, 2, 3] but got [1, 2, 3]
	at org.testng.Assert.fail(Assert.java:110)
	at org.testng.Assert.assertEquals(Assert.java:1825)
	at org.testng.Assert.assertEquals(Assert.java:1752)
	at com.facebook.presto.operator.TestOperatorStats.assertExpectedOperatorStats(TestOperatorStats.java:228)
	at com.facebook.presto.operator.TestDriverStats.assertExpectedDriverStats(TestDriverStats.java:111)
	at com.facebook.presto.operator.TestDriverStats.testJson(TestDriverStats.java:75)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:569)
	at org.testng.internal.invokers.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:135)
	at org.testng.internal.invokers.TestInvoker.invokeMethod(TestInvoker.java:673)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethod(TestInvoker.java:220)
	at org.testng.internal.invokers.MethodRunner.runInSequence(MethodRunner.java:50)
	at org.testng.internal.invokers.TestInvoker$MethodInvocationAgent.invoke(TestInvoker.java:945)
	at org.testng.internal.invokers.TestInvoker.invokeTestMethods(TestInvoker.java:193)
	at org.testng.internal.invokers.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:146)
	at org.testng.internal.invokers.TestMethodWorker.run(TestMethodWorker.java:128)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:840)

[INFO] 
[INFO] Results:
[INFO] 
Error:  Failures: 
Error:    TestDriverStats.testJson:75->assertExpectedDriverStats:111 Sets differ: expected [1, 2, 3] but got [1, 2, 3]
Error:    TestOperatorStats.testAddMixedStartingWithMergeable:282 Sets differ: expected [1, 2, 3] but got [1, 2, 3]
Error:    TestOperatorStats.testAddMixedStartingWithNonMergeable:388 Sets differ: expected [1, 2, 3] but got [1, 2, 3]
Error:    TestOperatorStats.testAddNonMergeable:438 Sets differ: expected [1, 2, 3] but got [1, 2, 3]
Error:    TestOperatorStats.testJson:173->assertExpectedOperatorStats:228 Sets differ: expected [1, 2, 3] but got [1, 2, 3]
Error:    TestTaskStats.testJson:97->assertExpectedTaskStats:144 Sets differ: expected [1, 2, 3] but got [1, 2, 3]
[INFO] 
Error:  Tests run: 9745, Failures: 6, Errors: 0, Skipped: 1

```

It appears that `DynamicFilterStats.copyOf(...)` and `DynamicFilterStats.mergeWith(...)` do not provide isolation from the original `DynamicFilterStats` object. As a result, the following code in the test directly affects the shared object `TEST_DYNAMIC_FILTER_STATS_1`, which is supposed to be immutable: 

```
        DynamicFilterStats expectedDynamicFilterStats = DynamicFilterStats.copyOf(TEST_DYNAMIC_FILTER_STATS_1);
        expectedDynamicFilterStats.mergeWith(TEST_DYNAMIC_FILTER_STATS_2);
```

This may cause issues when test cases are executed concurrently.

--- a/presto-main-base/src/main/java/com/facebook/presto/operator/DynamicFilterStats.java
+++ b/presto-main-base/src/main/java/com/facebook/presto/operator/DynamicFilterStats.java
@@ -19,7 +19,9 @@ import com.facebook.drift.annotations.ThriftStruct;
 import com.facebook.presto.spi.plan.PlanNodeId;
 import com.fasterxml.jackson.annotation.JsonCreator;
 import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.annotations.VisibleForTesting;
 
+import java.util.HashSet;
 import java.util.Set;
 
 import static java.util.Objects.requireNonNull;
@@ -45,10 +47,11 @@ public class DynamicFilterStats
         this.producerNodeIds = requireNonNull(producerNodeIds, "producerNodeIds is null");
     }
 
+    @VisibleForTesting
     public static DynamicFilterStats copyOf(DynamicFilterStats dynamicFilterStats)
     {
         requireNonNull(dynamicFilterStats, "dynamicFilterStats is null");
-        return new DynamicFilterStats(dynamicFilterStats.getProducerNodeIds());
+        return new DynamicFilterStats(new HashSet<>(dynamicFilterStats.getProducerNodeIds()));
     }
 
     public void mergeWith(DynamicFilterStats other)

--- a/presto-main-base/src/test/java/com/facebook/presto/operator/TestOperatorStats.java
+++ b/presto-main-base/src/test/java/com/facebook/presto/operator/TestOperatorStats.java
@@ -385,7 +385,7 @@ public class TestOperatorStats
 
         DynamicFilterStats expectedDynamicFilterStats = DynamicFilterStats.copyOf(TEST_DYNAMIC_FILTER_STATS_1);
         expectedDynamicFilterStats.mergeWith(TEST_DYNAMIC_FILTER_STATS_2);
-        assertEquals(actual.getDynamicFilterStats().getProducerNodeIds(), TEST_DYNAMIC_FILTER_STATS_1.getProducerNodeIds());
+        assertEquals(actual.getDynamicFilterStats().getProducerNodeIds(), expectedDynamicFilterStats.getProducerNodeIds());
     }
 
     @Test

